torch/cuda/amp/grad_scaler.py

```diff
@@ -329,7 +329,16 @@ def step(self, optimizer, *args, **kwargs):
             # The contract with custom optimizers is that their step() should accept an additional,
             # optional grad_scaler kwarg.  We append self to the kwargs so the custom optimizer has full information:
             # it can query its own state, invoke unscale_ on itself, etc
-            retval = optimizer.step(*args, **dict(kwargs, grad_scaler=self))
+            scaler = self._get_scale_async()
+            found_inf = cast(
+                torch.Tensor,
+                sum([
+                    t.to(scaler.device, non_blocking=True) for t in self._check_inf_per_device(optimizer).values()
+                ])
+            )
+            optimizer.grad_scale = scaler
+            optimizer.found_inf = found_inf
+            retval = optimizer.step(*args)
```

Comment:
@albanD may I get your feedback on the argument changes including type hints in adam.py and this draft change which doesn't add the new keyword argument of `grad_scaler` to `optimizer.step`?
even if this looks okay, this would be a breaking change in some sense so I'd need to check if the passed optimizer has the argument for a while